lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp

```diff
@@ -13,155 +13,219 @@ namespace PlatformAgnostic
+        int t = ymd->time;
+        t /= DateTimeTicks_PerSecond; // discard ms
+        time->tm_sec = t % 60;
+        t /= 60;
+        time->tm_min = t % 60;
+        t /= 60;
+        time->tm_hour = t;
+
+        // mktime etc. broken-out time accepts 1900 as a start year while epoch is 1970
+        // temporarily add a calendar day for leap pass
+        bool leap_year = IsLeap(time->tm_year);
+        *leap_added = false;
+        if (ymd->yday == 60 && leap_year) {
+            time->tm_mday++;
+            *leap_added = true;
+        }
+    }
 
-        #if defined(WCHAR_IS_CHAR16_T)
-            for(int i = 0; i < standardZoneNameLength; i++) {
-                standardZoneName[i] = (char16_t)ltime.tm_zone[i];
-            }
-        #elif defined(WCHAR_IS_WCHAR_T)
-            mbstowcs( (wchar_t*) standardZoneName, ltime.tm_zone,
-                      sizeof(standardZoneName) / sizeof(WCHAR));
-        #else
-        #error "WCHAR should be either wchar_t or char16_t"
-        #endif
+    static void TM_TO_YMD(const struct tm *time, YMD *ymd, const bool leap_added, const int base_year)
+    {
+        ymd->year = UpdateToYMDYear(base_year, time);
+        ymd->mon = time->tm_mon;
+        ymd->mday = time->tm_mday;
 
-            standardZoneName[standardZoneNameLength] = (WCHAR)0;
-            lastTimeZoneUpdateTickCount = tickCount;
+        ymd->time = (time->tm_hour * DateTimeTicks_PerHour)
+                  + (time->tm_min * DateTimeTicks_PerMinute)
+                  + (time->tm_sec * DateTimeTicks_PerSecond);
+
+        // mktime etc. broken-out time accepts 1900 as a start year while epoch is 1970
+        // minus the previously added calendar day (see YMD_TO_TM)
+        if (leap_added)
+        {
+            AssertMsg(ymd->mday >= 0, "Day of month can't be a negative number");
+            if (ymd->mday == 0)
+            {
+                ymd->mday = 29;
+                ymd->mon = 1;
+            }
+            else
+            {
+                ymd->mday--;
+            }
         }
     }
 
-    // DateTimeHelper ******
-    static inline void TM_TO_SYSTIME(struct tm *time, SYSTEMTIME *sysTime)
+    static void CopyTimeZoneName(WCHAR *wstr, size_t *length, const char *tm_zone)
     {
-        sysTime->wYear = time->tm_year + 1900;
-        sysTime->wMonth = time->tm_mon + 1;
-        sysTime->wDayOfWeek = time->tm_wday;
-        sysTime->wDay = time->tm_mday;
-        sysTime->wHour = time->tm_hour;
-        sysTime->wMinute = time->tm_min;
-
-        // C99 tm_sec value is between 0 and 60. (1 leap second.)
-        // Discard leap second.
-        // In any case, this is a representation of a wallclock time.
-        // It can jump forwards and backwards.
-
-        sysTime->wSecond = time->tm_sec % 60;
+        *length = strlen(tm_zone);
+
+#if defined(WCHAR_IS_CHAR16_T)
+        for(int i = 0; i < *length; i++) {
+            wstr[i] = (char16_t)tm_zone[i];
+        }
+#elif defined(WCHAR_IS_WCHAR_T)
```

Comment:
Can we get rid of this? Should be safe to assume that on non-Win32 platforms, wchar is char16?